cli/src/semgrep/console.py

```diff
@@ -25,18 +27,81 @@ class Title:
 
     def __rich__(self) -> RenderableType:
         if self.order == 1:
-            return Padding(Panel(self.text, expand=False, box=box.SQUARE), (2, 0, 1, 0))
+            return Padding(Panel(self.text, expand=False, box=box.SQUARE), (2, 0, 0, 0))
         elif self.order == 2:
-            return Padding(Text(self.text.upper(), style="underline"), (0, 2))
+            return Padding(Text(self.text.upper(), style="underline"), (1, 0, 0, 2))
         elif self.order == 3:
-            return Text("⮕ " + self.text.upper())
+            return Text("  ⮕ " + self.text.upper())
         else:
             raise ValueError(f"Title order must be 1, 2, or 3, not {self.order}")
 
 
+class AutoIndentingConsole(Console):
+    """This custom console keeps track of the last title printed and automatically indents the next line according to what level title we're under."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        self.active_title: Optional[Title] = None
+        super().__init__(*args, **kwargs)
+
+    @property
+    def auto_indent_size(self) -> int:
+        if self.active_title is None:
+            return 0
+
+        if self.active_title.order == 1:
+            return 2
+        elif self.active_title.order == 2:
+            return 2
+        elif self.active_title.order == 3:
+            return 5
+        else:
+            return 0
+
+    def line(self, *_: Any, **__: Any) -> None:
+        """Whenever rich internally tries to print a blank line, we ignore it.
+
+        This way we have greater control over spacing."""
+
+    def reset_title(self) -> None:
+        self.active_title = None
+
+    @staticmethod
+    def extract_title(current: Optional[RenderableType]) -> Optional[Title]:
+        while isinstance(current, Title) or hasattr(current, "renderable"):
+            if isinstance(current, Title):
+                return current
+            current = getattr(current, "renderable", None)
+
+        return None
+
+    def print(self, *args: Any, deindent: int = 0, **kwargs: Any) -> None:
+        indent = 0
+
+        for arg in args:
+            title = self.extract_title(arg)
+            if title is not None:
+                self.active_title = title
```

Comment:
what does setting self.active_title do?

Reply:
Looks like this is handling printing titles but not sure how. Would it also be more correct to special case printing strings so console correctly handles any other future objects passed in (aside from Title class)

Reply:
Oh self.active_title is var we created nvm